bash_completion

```diff
@@ -2208,6 +2208,18 @@ _comp_command_offset()
     # rewrite current completion context before invoking
     # actual command completion
 
+    # convert the new first-word position $1 to the index in COMP_WORDS.
+    if [[ -v cword && -v words ]]; then
+        local reassembled_offset=$1 i=0 j
+        for ((j = 0; j < reassembled_offset; j++)); do
+            local word=${words[j]}
+            while [[ $word && i -lt ${#COMP_WORDS[@]} ]]; do
+                word=${word#*"${COMP_WORDS[i++]}"}
+            done
```

Comment:
I'm sorry and afraid that I'm having a bit of a problem understanding the intent and workings here. So we remove things found in `COMP_WORDS` from each `word` until the `word` becomes empty or we run out of `COMP_WORDS`. But as we treat the thing to be removed as a `word` prefix, wouldn't we run into problems / false positives with let's say the sequence `sudo su do`?

Could we add some unit tests for this, to illustrate how the changed code works?

Reply:
> the intent and workings here.

We here convert the index in `words` to the index in `COMP_WORDS`. The array `words` is constructed by the function `_comp__reassemble_words` (originally `__reassemble_comp_words_by_ref`). [The function `_comp__reassemble_words` combines multiple `COMP_WORDS[*]` into an element of `words`](https://github.com/scop/bash-completion/blob/e9bf32c3ade5e4f72321638eb3928d66bc0ec1bd/bash_completion#L526), so we need to decomposes `words[j]` by referencing the original `COMP_WORDS[i]`. For the approach `${word#*"${COMOP_WORDS[i++]}"}`, I borrowed [the existing strategy for `line` used in `__comp__reassemble_words`](https://github.com/scop/bash-completion/blob/e9bf32c3ade5e4f72321638eb3928d66bc0ec1bd/bash_completion#L528).

> But as we treat the thing to be removed as a `word` prefix, wouldn't we run into problems / false positives with let's say the sequence `sudo su do`?

The assumption is that `words` are constructed from `COMP_WORDS` using `_comp__reassemble_words`, so that we always have a corresponding substring in `words[j]` for any `COMP_WORDS[i]`. In this case, even with `sudo su do`, the correspondence between `words` and `COMP_WORDS` are kept so unrelated words will not be mixed.

I'll later add tests.

Reply:
I can't shake the feeling that for some input we'd lose track of things here, because we're just repeatedly removing a prefix from the  `words` entry apparently without knowledge whether it ended up there due to the earlier reassembly, or if it was there already. Somehow this is quite confusing to me, but then again so is the `_comp__reassemble_words` implementation despite the commentary in it.

I guess seeing some tests will help. Extracting the "reverse word index reassembly" block that was added here to a separate function could be good for testability.

Reply:
I separated a function `_comp__find_original_word`. The name of the function can be adjusted.

I also added two tests, but I'm not sure what the concern is, so they might not cover the situation that you concern about.